drivers/can/can_nxp_s32_canxl.c

```diff
@@ -0,0 +1,1097 @@
+
+	Canexcel_Ip_SetRxIndividualMask(config->instance, mb_indx,
+						data->rx_cbs[alloc].rx_info.frame, mask);
+
+	Canexcel_Ip_ConfigRx(config->instance, mb_indx, filter->id,
+					&data->rx_cbs[alloc].rx_info);
+
+	Canexcel_Ip_ReceiveFD(config->instance, mb_indx, &data->rx_msg[alloc], FALSE);
+
+	Canexcel_Ip_ExitFreezeMode(config->instance);
+
+unlock:
+	k_mutex_unlock(&data->rx_mutex);
+
+	return alloc;
+}
+
+static int can_nxp_s32_send(const struct device *dev,
+				const struct can_frame *frame,
+				k_timeout_t timeout,
+				can_tx_callback_t callback, void *user_data)
+{
+	const struct can_nxp_s32_config *config = dev->config;
+	uint8_t data_length = can_dlc_to_bytes(frame->dlc);
+	struct can_nxp_s32_data *data = dev->data;
+	Canexcel_Ip_StatusType status;
+	enum can_state state;
+	int alloc, mb_indx;
+
+	__ASSERT_NO_MSG(callback != NULL);
+
+#ifdef CONFIG_CAN_FD_MODE
+	if ((frame->flags & ~(CAN_FRAME_IDE | CAN_FRAME_FDF | CAN_FRAME_BRS)) != 0) {
+		LOG_ERR("unsupported CAN frame flags 0x%02x", frame->flags);
+		return -ENOTSUP;
+	}
+
+	if ((frame->flags & CAN_FRAME_FDF) != 0 &&
+			(config->base_sic->BCFG2 & CANXL_SIC_BCFG2_FDEN_MASK) == 0) {
+		LOG_ERR("CAN-FD format not supported in non-FD mode");
+		return -ENOTSUP;
+	}
+
+	if ((frame->flags & CAN_FRAME_BRS) != 0 &&
+			~(config->base_sic->BCFG1 & CANXL_SIC_BCFG1_FDRSDIS_MASK) == 0) {
+		LOG_ERR("CAN-FD BRS not supported in non-FD mode");
+		return -ENOTSUP;
+	}
+#else
+	if ((frame->flags & ~CAN_FRAME_IDE) != 0) {
+		LOG_ERR("unsupported CAN frame flags 0x%02x", frame->flags);
+		return -ENOTSUP;
+	}
+#endif
+
+	if (data_length > sizeof(frame->data)) {
+		LOG_ERR("data length (%d) > max frame data length (%d)",
+			data_length, sizeof(frame->data));
+		return -EINVAL;
+	}
+
+	if ((frame->flags & CAN_FRAME_FDF) == 0) {
+		if (frame->dlc > CAN_MAX_DLC) {
+			LOG_ERR("DLC of %d for non-FD format frame", frame->dlc);
+			return -EINVAL;
+		}
+#ifdef CONFIG_CAN_FD_MODE
+	} else {
+		if (frame->dlc > CANFD_MAX_DLC) {
+			LOG_ERR("DLC of %d for CAN-FD format frame", frame->dlc);
+			return -EINVAL;
+		}
+#endif
+	}
+
+	if (!data->started) {
+		return -ENETDOWN;
+	}
+
+	can_nxp_s32_get_state(dev, &state, NULL);
+	if (state == CAN_STATE_BUS_OFF) {
+		LOG_ERR("Transmit failed, bus-off");
+		return -ENETUNREACH;
+	}
+
+	if (k_sem_take(&data->tx_allocs_sem, timeout) != 0) {
+		return -EAGAIN;
+	}
+
+	for (alloc = 0; alloc < CONFIG_CAN_NXP_S32_MAX_TX; alloc++) {
+		if (!atomic_test_and_set_bit(data->tx_allocs, alloc)) {
+			break;
+		}
+	}
+
+	data->tx_cbs[alloc].function = callback;
+	data->tx_cbs[alloc].arg = user_data;
+	mb_indx = ALLOC_IDX_TO_TXMB_IDX(alloc);
+	data->tx_cbs[alloc].tx_info = (Canexcel_Ip_DataInfoType) {
+		.frame = !!(frame->flags & CAN_FRAME_FDF) ?
+				CANEXCEL_FD_FRAME : CANEXCEL_CLASIC_FRAME,
+		.enable_brs = !!(frame->flags & CAN_FRAME_BRS) ? TRUE : FALSE,
+		.idType = !!(frame->flags & CAN_FRAME_IDE) ?
+				CANEXCEL_MSG_ID_EXT : CANEXCEL_MSG_ID_STD,
```

Comment:
I removed the setting ` .retransmission = 0 ` in `can_send()`https://github.com/zephyrproject-rtos/zephyr/pull/52450#discussion_r1178311886